File: src/ontology/procedural.rs
//! Procedural Ontology Module
//!
//! Procedural ontology nodes describe workflows, execution steps, decision points,
//! failure modes, and playbooks.

use super::{normalize_alias, OntologyGid};
use serde::{Deserialize, Serialize};

/// Procedural element types for the procedural layer
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ProceduralElementType {
    Workflow,
    WorkflowStep,
    DecisionPoint,
    FailureMode,
    PlaybookStep,
}

impl ProceduralElementType {
    pub fn as_str(&self) -> &'static str {
        match self {
            ProceduralElementType::Workflow => "workflow",
            ProceduralElementType::WorkflowStep => "workflow_step",
            ProceduralElementType::DecisionPoint => "decision_point",
            ProceduralElementType::FailureMode => "failure_mode",
            ProceduralElementType::PlaybookStep => "playbook_step",
        }
    }

    #[allow(clippy::should_implement_trait)]
    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "workflow" => Some(Self::Workflow),
            "workflow_step" => Some(Self::WorkflowStep),
            "decision_point" => Some(Self::DecisionPoint),
            "failure_mode" => Some(Self::FailureMode),
            "playbook_step" => Some(Self::PlaybookStep),
            _ => None,
        }
    }
}

/// Procedural relationship types
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ProceduralRelationshipType {
    HasStep,
    NextStep,
    BranchesTo,
    ImplementedBy,
    HasFailureMode,
    HandledByPlaybook,
}

impl ProceduralRelationshipType {
    pub fn as_str(&self) -> &'static str {
        match self {
            ProceduralRelationshipType::HasStep => "has_step",
            ProceduralRelationshipType::NextStep => "next_step",
            ProceduralRelationshipType::BranchesTo => "branches_to",
            ProceduralRelationshipType::ImplementedBy => "implemented_by",
            ProceduralRelationshipType::HasFailureMode => "has_failure_mode",
            ProceduralRelationshipType::HandledByPlaybook => "handled_by_playbook",
        }
    }

    #[allow(clippy::should_implement_trait)]
    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "has_step" => Some(Self::HasStep),
            "next_step" => Some(Self::NextStep),
            "branches_to" => Some(Self::BranchesTo),
            "implemented_by" => Some(Self::ImplementedBy),
            "has_failure_mode" => Some(Self::HasFailureMode),
            "handled_by_playbook" => Some(Self::HandledByPlaybook),
            _ => None,
        }
    }
}

/// Metadata contract for workflow nodes
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct WorkflowMetadata {
    pub gid: String,
    pub ontology: String,
    pub ontology_layer: String,
    #[serde(default)]
    pub aliases: Vec<String>,
    pub description: String,
    #[serde(default)]
    pub entry_points: Vec<String>,
    #[serde(default)]
    pub step_count: Option<usize>,
    #[serde(default)]
    pub source: Option<String>,
    #[serde(default)]
    pub valid_from: Option<String>,
    #[serde(default)]
    pub valid_until: Option<String>,
    #[serde(default)]
    pub stale: bool,
    #[serde(default)]
    pub stale_reason: Option<String>,
    #[serde(default)]
    pub last_seen_at: Option<String>,
}

impl WorkflowMetadata {
    pub fn new(env: &str, scope: &str, workflow_id: &str, name: &str, description: &str) -> Self {
        let gid = OntologyGid {
            env: env.to_string(),
            scope: scope.to_string(),
            ontology_type: "workflow".to_string(),
            id: workflow_id.to_string(),
            version: "v1".to_string(),
        }
        .to_string();

        Self {
            gid,
            ontology: "procedural".to_string(),
            ontology_layer: "procedural".to_string(),
            aliases: vec![normalize_alias(name)],
            description: description.to_string(),
            entry_points: Vec::new(),
            step_count: None,
            source: None,
            valid_from: None,
            valid_until: None,
            stale: false,
            stale_reason: None,
            last_seen_at: None,
        }
    }

    pub fn with_entry_points(mut self, entry_points: Vec<String>) -> Self {
        self.entry_points = entry_points;
        self
    }

    pub fn with_step_count(mut self, count: usize) -> Self {
        self.step_count = Some(count);
        self
    }

    pub fn with_source(mut self, source: &str) -> Self {
        self.source = Some(source.to_string());
        self
    }
}

/// Metadata contract for workflow step nodes
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct WorkflowStepMetadata {
    pub gid: String,
    pub ontology: String,
    pub ontology_layer: String,
    pub workflow_gid: String,
    pub order: usize,
    #[serde(default)]
    pub aliases: Vec<String>,
    pub description: String,
    #[serde(default)]
    pub code_refs: Vec<String>,
    #[serde(default)]
    pub failure_modes: Vec<String>,
    #[serde(default)]
    pub source: Option<String>,
    #[serde(default)]
    pub stale: bool,
    #[serde(default)]
    pub stale_reason: Option<String>,
    #[serde(default)]
    pub last_seen_at: Option<String>,
}

impl WorkflowStepMetadata {
    pub fn new(
        env: &str,
        scope: &str,
        workflow_id: &str,
        step_id: &str,
        order: usize,
        description: &str,
    ) -> Self {
        let gid = OntologyGid {
            env: env.to_string(),
            scope: scope.to_string(),
            ontology_type: "workflow_step".to_string(),
            id: step_id.to_string(),
            version: "v1".to_string(),
        }
        .to_string();

        Self {
            gid,
            ontology: "procedural".to_string(),
            ontology_layer: "procedural".to_string(),
            workflow_gid: format!("{}:{}:{}:{}:{}", env, scope, "workflow", workflow_id, "v1"),
            order,
            aliases: Vec::new(),
            description: description.to_string(),
            code_refs: Vec::new(),
            failure_modes: Vec::new(),
            source: None,
            stale: false,
            stale_reason: None,
            last_seen_at: None,
        }
    }

    pub fn with_code_refs(mut self, refs: Vec<String>) -> Self {
        self.code_refs = refs;
        self
    }

    pub fn with_failure_modes(mut self, modes: Vec<String>) -> Self {
        self.failure_modes = modes;
        self
    }
}

/// Workflow node structure for graph insertion
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowNode {
    pub gid: String,
    pub name: String,
    pub element_type: String,
    pub aliases: Vec<String>,
    pub description: String,
    pub env: String,
    pub metadata: WorkflowMetadata,
}

impl WorkflowNode {
    pub fn new(env: &str, scope: &str, workflow_id: &str, name: &str, description: &str) -> Self {
        let metadata = WorkflowMetadata::new(env, scope, workflow_id, name, description);

        Self {
            gid: metadata.gid.clone(),
            name: name.to_string(),
            element_type: "workflow".to_string(),
            aliases: metadata.aliases.clone(),
            description: description.to_string(),
            env: env.to_string(),
            metadata,
        }
    }

    pub fn with_entry_points(mut self, entry_points: Vec<String>) -> Self {
        self.metadata = self.metadata.with_entry_points(entry_points);
        self
    }

    pub fn with_step_count(mut self, count: usize) -> Self {
        self.metadata = self.metadata.with_step_count(count);
        self
    }

    pub fn with_source(mut self, source: &str) -> Self {
        self.metadata = self.metadata.with_source(source);
        self
    }
}

/// Workflow step node structure
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkflowStepNode {
    pub gid: String,
    pub name: String,
    pub element_type: String,
    pub workflow_gid: String,
    pub order: usize,
    pub description: String,
    pub env: String,
    pub metadata: WorkflowStepMetadata,
}

impl WorkflowStepNode {
    pub fn new(
        env: &str,
        scope: &str,
        workflow_id: &str,
        step_id: &str,
        name: &str,
        order: usize,
        description: &str,
    ) -> Self {
        let metadata =
            WorkflowStepMetadata::new(env, scope, workflow_id, step_id, order, description);

        Self {
            gid: metadata.gid.clone(),
            name: name.to_string(),
            element_type: "workflow_step".to_string(),
            workflow_gid: metadata.workflow_gid.clone(),
            order,
            description: description.to_string(),
            env: env.to_string(),
            metadata,
        }
    }

    pub fn with_code_refs(mut self, refs: Vec<String>) -> Self {
        self.metadata = self.metadata.with_code_refs(refs);
        self
    }

    pub fn with_failure_modes(mut self, modes: Vec<String>) -> Self {
        self.metadata = self.metadata.with_failure_modes(modes);
        self
    }
}

/// Failure mode node
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FailureModeNode {
    pub gid: String,
    pub name: String,
    pub element_type: String,
    pub description: String,
    pub env: String,
    pub metadata: FailureModeMetadata,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct FailureModeMetadata {
    pub gid: String,
    pub ontology: String,
    pub ontology_layer: String,
    #[serde(default)]
    pub aliases: Vec<String>,
    pub description: String,
    #[serde(default)]
    pub handled_by: Vec<String>,
    #[serde(default)]
    pub source: Option<String>,
    #[serde(default)]
    pub stale: bool,
    #[serde(default)]
    pub stale_reason: Option<String>,
    #[serde(default)]
    pub last_seen_at: Option<String>,
}

impl FailureModeNode {
    pub fn new(env: &str, scope: &str, failure_id: &str, name: &str, description: &str) -> Self {
        let gid = OntologyGid {
            env: env.to_string(),
            scope: scope.to_string(),
            ontology_type: "failure_mode".to_string(),
            id: failure_id.to_string(),
            version: "v1".to_string(),
        }
        .to_string();

        let metadata = FailureModeMetadata {
            gid: gid.clone(),
            ontology: "procedural".to_string(),
            ontology_layer: "procedural".to_string(),
            aliases: vec![normalize_alias(name)],
            description: description.to_string(),
            handled_by: Vec::new(),
            source: None,
            stale: false,
            stale_reason: None,
            last_seen_at: None,
        };

        Self {
            gid,
            name: name.to_string(),
            element_type: "failure_mode".to_string(),
            description: description.to_string(),
            env: env.to_string(),
            metadata,
        }
    }

    pub fn with_handled_by(mut self, playbooks: Vec<String>) -> Self {
        self.metadata.handled_by = playbooks;
        self
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_procedural_element_type_as_str() {
        assert_eq!(ProceduralElementType::Workflow.as_str(), "workflow");
        assert_eq!(
            ProceduralElementType::WorkflowStep.as_str(),
            "workflow_step"
        );
        assert_eq!(ProceduralElementType::FailureMode.as_str(), "failure_mode");
    }

    #[test]
    fn test_procedural_element_type_from_str() {
        assert_eq!(
            ProceduralElementType::from_str("workflow"),
            Some(ProceduralElementType::Workflow)
        );
        assert_eq!(
            ProceduralElementType::from_str("workflow_step"),
            Some(ProceduralElementType::WorkflowStep)
        );
        assert_eq!(ProceduralElementType::from_str("unknown"), None);
    }

    #[test]
    fn test_procedural_relationship_type_roundtrip() {
        assert_eq!(ProceduralRelationshipType::HasStep.as_str(), "has_step");
        assert_eq!(
            ProceduralRelationshipType::from_str("has_step"),
            Some(ProceduralRelationshipType::HasStep)
        );
        assert_eq!(
            ProceduralRelationshipType::from_str("next_step"),
            Some(ProceduralRelationshipType::NextStep)
        );
    }

    #[test]
    fn test_workflow_metadata_new() {
        let meta = WorkflowMetadata::new(
            "local",
            "checkout-service",
            "checkout",
            "Checkout",
            "End-to-end customer checkout workflow",
        );
        assert_eq!(meta.gid, "local:checkout-service:workflow:checkout:v1");
        assert_eq!(meta.ontology, "procedural");
        assert_eq!(meta.ontology_layer, "procedural");
    }

    #[test]
    fn test_workflow_step_metadata_workflow_gid() {
        let meta = WorkflowStepMetadata::new(
            "local",
            "checkout-service",
            "checkout",
            "authorize_payment",
            2,
            "Authorize payment step",
        );
        assert_eq!(
            meta.workflow_gid,
            "local:checkout-service:workflow:checkout:v1"
        );
        assert_eq!(meta.order, 2);
    }

    #[test]
    fn test_workflow_node_creation() {
        let workflow = WorkflowNode::new(
            "local",
            "checkout-service",
            "checkout",
            "Checkout",
            "End-to-end customer checkout workflow",
        )
        .with_entry_points(vec!["src/checkout/handler.rs::create_order".to_string()])
        .with_step_count(5)
        .with_source("detected_from_call_graph");

        assert_eq!(workflow.gid, "local:checkout-service:workflow:checkout:v1");
        assert_eq!(workflow.name, "Checkout");
        assert_eq!(workflow.metadata.entry_points.len(), 1);
        assert_eq!(workflow.metadata.step_count, Some(5));
    }

    #[test]
    fn test_workflow_step_node_creation() {
        let step = WorkflowStepNode::new(
            "local",
            "checkout-service",
            "checkout",
            "authorize_payment",
            "Authorize Payment",
            2,
            "Authorize payment for the order",
        )
        .with_code_refs(vec!["src/payment/client.rs::authorize".to_string()])
        .with_failure_modes(vec![
            "payment_timeout".to_string(),
            "insufficient_funds".to_string(),
        ]);

        assert_eq!(
            step.gid,
            "local:checkout-service:workflow_step:authorize_payment:v1"
        );
        assert_eq!(step.name, "Authorize Payment");
        assert_eq!(step.order, 2);
        assert_eq!(step.metadata.code_refs.len(), 1);
        assert_eq!(step.metadata.failure_modes.len(), 2);
    }

    #[test]
    fn test_failure_mode_node_creation() {
        let failure = FailureModeNode::new(
            "local",
            "checkout-service",
            "payment_timeout",
            "Payment Timeout",
            "Payment authorization timed out",
        )
        .with_handled_by(vec![
            "local:checkout-service:playbook:payment_reconciliation:v1".to_string(),
        ]);

        assert_eq!(
            failure.gid,
            "local:checkout-service:failure_mode:payment_timeout:v1"
        );
        assert_eq!(failure.name, "Payment Timeout");
        assert_eq!(failure.metadata.handled_by.len(), 1);
    }
}

File: go/internal/ontology/procedural.go
package ontology

import (
	"encoding/json"
	"fmt"
	"strings"
	"unicode"
)

// Procedural element types for the procedural layer (parity with
// ProceduralElementType in the Rust reference).
const (
	TypeWorkflow      = "workflow"
	TypeWorkflowStep  = "workflow_step"
	TypeDecisionPoint = "decision_point"
	TypeFailureMode   = "failure_mode"
	TypePlaybookStep  = "playbook_step"
)

// Concept (domain layer) element types (parity with ConceptElementType).
const (
	TypeDomainEntity  = "domain_entity"
	TypeService       = "service"
	TypeAPIEndpoint   = "api_endpoint"
	TypeDataStore     = "data_store"
	TypeEnvironment   = "environment"
	TypeKnownIssue    = "known_issue"
	TypePlaybook      = "playbook"
	TypeTeamKnowledge = "team_knowledge"
)

// Procedural relationship types (parity with ProceduralRelationshipType).
const (
	RelHasStep           = "has_step"
	RelNextStep          = "next_step"
	RelBranchesTo        = "branches_to"
	RelImplementedBy     = "implemented_by"
	RelHasFailureMode    = "has_failure_mode"
	RelHandledByPlaybook = "handled_by_playbook"
)

// Concept relationship types (parity with ConceptRelationshipType).
const (
	RelOwnsConcept        = "owns_concept"
	RelImplementsConcept  = "implements_concept"
	RelExposesEndpoint    = "exposes_endpoint"
	RelReadsFrom          = "reads_from"
	RelWritesTo           = "writes_to"
	RelDocumentsConcept   = "documents_concept"
	RelHasKnownIssue      = "has_known_issue"
	RelResolvedByPlaybook = "resolved_by_playbook"
)

// ProceduralTypes is the set of procedural-layer element types.
var ProceduralTypes = map[string]bool{
	TypeWorkflow: true, TypeWorkflowStep: true, TypeDecisionPoint: true,
	TypeFailureMode: true, TypePlaybookStep: true,
}

// ConceptTypes is the set of domain-layer element types.
var ConceptTypes = map[string]bool{
	TypeDomainEntity: true, TypeService: true, TypeAPIEndpoint: true,
	TypeDataStore: true, TypeEnvironment: true, TypeKnownIssue: true,
	TypePlaybook: true, TypeTeamKnowledge: true,
}

// OntologyElements is every element type the ontology layer owns.
var OntologyElements = map[string]bool{
	TypeDomainEntity: true, TypeService: true, TypeAPIEndpoint: true,
	TypeDataStore: true, TypeEnvironment: true, TypeKnownIssue: true,
	TypePlaybook: true, TypeTeamKnowledge: true,
	TypeWorkflow: true, TypeWorkflowStep: true, TypeDecisionPoint: true,
	TypeFailureMode: true, TypePlaybookStep: true,
}

// IsProceduralType reports whether t belongs to the procedural layer.
func IsProceduralType(t string) bool { return ProceduralTypes[t] }

// GID is the stable global id of an ontology node,
// "env:scope:ontology_type:id:version" (parity with OntologyGid).
type GID struct {
	Env          string
	Scope        string
	OntologyType string
	ID           string
	Version      string
}

// ParseGid parses a GID string like "local:checkout-service:domain_entity:refund:v1".
func ParseGid(s string) (GID, bool) {
	parts := strings.Split(s, ":")
	if len(parts) != 5 {
		return GID{}, false
	}
	return GID{Env: parts[0], Scope: parts[1], OntologyType: parts[2], ID: parts[3], Version: parts[4]}, true
}

// Format renders the canonical "env:scope:type:id:version" form.
func (g GID) Format() string {
	return fmt.Sprintf("%s:%s:%s:%s:%s", g.Env, g.Scope, g.OntologyType, g.ID, g.Version)
}

// NewGid builds a version-1 GID.
func NewGid(env, scope, ontologyType, id string) GID {
	return GID{Env: env, Scope: scope, OntologyType: ontologyType, ID: id, Version: "v1"}
}

// NormalizeAlias normalizes an alias: lowercase, trimmed, keeping only
// alphanumeric characters, spaces, hyphens and underscores (parity with
// normalize_alias).
func NormalizeAlias(alias string) string {
	alias = strings.ToLower(alias)
	alias = strings.TrimSpace(alias)
	var b strings.Builder
	for _, c := range alias {
		if isAliasRune(c) {
			b.WriteRune(c)
		}
	}
	return b.String()
}

func isAliasRune(c rune) bool {
	switch {
	case c >= 'a' && c <= 'z', c >= '0' && c <= '9':
		return true
	case c == ' ' || c == '-' || c == '_':
		return true
	default:
		return unicode.IsLetter(c) || unicode.IsDigit(c)
	}
}

// NormalizeAliases normalizes each alias in turn.
func NormalizeAliases(aliases []string) []string {
	out := make([]string, len(aliases))
	for i, a := range aliases {
		out[i] = NormalizeAlias(a)
	}
	return out
}

// ConceptMetadata is the metadata contract for concept ontology nodes.
// JSON tags mirror the Rust serde field names so stored metadata stays
// byte-comparable with the Rust engine's.
type ConceptMetadata struct {
	GID           string   `json:"gid"`
	Ontology      string   `json:"ontology"`
	OntologyLayer string   `json:"ontology_layer"`
	Aliases       []string `json:"aliases,omitempty"`
	Description   string   `json:"description"`
	Source        *string  `json:"source,omitempty"`
	ValidFrom     *string  `json:"valid_from,omitempty"`
	ValidUntil    *string  `json:"valid_until,omitempty"`
	OwnedBy       []string `json:"owned_by,omitempty"`
	CodeRefs      []string `json:"code_refs,omitempty"`
	Docs          []string `json:"docs,omitempty"`
	Stale         bool     `json:"stale"`
	StaleReason   *string  `json:"stale_reason,omitempty"`
	LastSeenAt    *string  `json:"last_seen_at,omitempty"`
}

// ConceptNode is a concept-layer node for graph insertion.
type ConceptNode struct {
	GID         string          `json:"gid"`
	Name        string          `json:"name"`
	ElementType string          `json:"element_type"`
	Aliases     []string        `json:"aliases"`
	Description string          `json:"description"`
	Env         string          `json:"env"`
	Metadata    ConceptMetadata `json:"metadata"`
}

// NewConceptNode builds a concept node with a name-derived seed alias
// (parity with ConceptMetadata::new).
func NewConceptNode(env, scope, elementType, id, name, description string) ConceptNode {
	g := NewGid(env, scope, elementType, id)
	seed := []string{NormalizeAlias(name)}
	return ConceptNode{
		GID:         g.Format(),
		Name:        name,
		ElementType: elementType,
		Aliases:     seed,
		Description: description,
		Env:         env,
		Metadata: ConceptMetadata{
			GID:           g.Format(),
			Ontology:      "concept",
			OntologyLayer: "domain",
			Aliases:       seed,
			Description:   description,
		},
	}
}

// WithAliases appends normalized aliases to the node (parity with with_aliases).
func (n *ConceptNode) WithAliases(aliases []string) {
	n.Aliases = append(n.Aliases, NormalizeAliases(aliases)...)
	n.Metadata.Aliases = n.Aliases
}

// WithSource records the YAML source path in the metadata.
func (n *ConceptNode) WithSource(source string) {
	n.Metadata.Source = &source
}

// WorkflowMetadata is the metadata contract for workflow nodes.
type WorkflowMetadata struct {
	GID           string   `json:"gid"`
	Ontology      string   `json:"ontology"`
	OntologyLayer string   `json:"ontology_layer"`
	Aliases       []string `json:"aliases,omitempty"`
	Description   string   `json:"description"`
	EntryPoints   []string `json:"entry_points,omitempty"`
	StepCount     *int     `json:"step_count,omitempty"`
	Source        *string  `json:"source,omitempty"`
	ValidFrom     *string  `json:"valid_from,omitempty"`
	ValidUntil    *string  `json:"valid_until,omitempty"`
	Stale         bool     `json:"stale"`
	StaleReason   *string  `json:"stale_reason,omitempty"`
	LastSeenAt    *string  `json:"last_seen_at,omitempty"`
}

// WorkflowNode is a workflow node for graph insertion.
type WorkflowNode struct {
	GID         string           `json:"gid"`
	Name        string           `json:"name"`
	ElementType string           `json:"element_type"`
	Aliases     []string         `json:"aliases"`
	Description string           `json:"description"`
	Env         string           `json:"env"`
	Metadata    WorkflowMetadata `json:"metadata"`
}

// NewWorkflowNode builds a workflow node (parity with WorkflowNode::new).
func NewWorkflowNode(env, scope, workflowID, name, description string) WorkflowNode {
	g := NewGid(env, scope, TypeWorkflow, workflowID).Format()
	seed := []string{NormalizeAlias(name)}
	return WorkflowNode{
		GID:         g,
		Name:        name,
		ElementType: TypeWorkflow,
		Aliases:     seed,
		Description: description,
		Env:         env,
		Metadata: WorkflowMetadata{
			GID:           g,
			Ontology:      "procedural",
			OntologyLayer: "procedural",
			Aliases:       seed,
			Description:   description,
		},
	}
}

// WorkflowStepMetadata is the metadata contract for workflow step nodes.
type WorkflowStepMetadata struct {
	GID           string   `json:"gid"`
	Ontology      string   `json:"ontology"`
	OntologyLayer string   `json:"ontology_layer"`
	WorkflowGid   string   `json:"workflow_gid"`
	Order         int      `json:"order"`
	Aliases       []string `json:"aliases,omitempty"`
	Description   string   `json:"description"`
	CodeRefs      []string `json:"code_refs,omitempty"`
	FailureModes  []string `json:"failure_modes,omitempty"`
	FeatureIDs    []string `json:"feature_ids,omitempty"`
	UserStoryIDs  []string `json:"user_story_ids,omitempty"`
	Source        *string  `json:"source,omitempty"`
	Stale         bool     `json:"stale"`
	StaleReason   *string  `json:"stale_reason,omitempty"`
	LastSeenAt    *string  `json:"last_seen_at,omitempty"`
}

// WorkflowStepNode is a workflow step node for graph insertion.
type WorkflowStepNode struct {
	GID         string               `json:"gid"`
	Name        string               `json:"name"`
	ElementType string               `json:"element_type"`
	Aliases     []string             `json:"aliases"`
	WorkflowGid string               `json:"workflow_gid"`
	Order       int                  `json:"order"`
	Description string               `json:"description"`
	Env         string               `json:"env"`
	Metadata    WorkflowStepMetadata `json:"metadata"`
}

// NewWorkflowStepNode builds a step node with a name-derived seed alias
// (parity with WorkflowStepNode::new).
func NewWorkflowStepNode(env, scope, workflowID, stepID, name string, order int, description string) WorkflowStepNode {
	g := NewGid(env, scope, TypeWorkflowStep, stepID).Format()
	wg := NewGid(env, scope, TypeWorkflow, workflowID).Format()
	seed := []string{NormalizeAlias(name)}
	return WorkflowStepNode{
		GID:         g,
		Name:        name,
		ElementType: TypeWorkflowStep,
		Aliases:     seed,
		WorkflowGid: wg,
		Order:       order,
		Description: description,
		Env:         env,
		Metadata: WorkflowStepMetadata{
			GID:           g,
			Ontology:      "procedural",
			OntologyLayer: "procedural",
			WorkflowGid:   wg,
			Order:         order,
			Aliases:       seed,
			Description:   description,
		},
	}
}

// WithAliases appends normalized aliases to the workflow node (parity with
// with_aliases).
func (n *WorkflowNode) WithAliases(aliases []string) {
	n.Metadata.Aliases = append(n.Metadata.Aliases, NormalizeAliases(aliases)...)
	n.Aliases = n.Metadata.Aliases
}

// WithAliases merges YAML-declared aliases on top of the name-derived seed
// (FR-HEA-01 parity: previously parsed but never applied in the Rust loader).
func (n *WorkflowStepNode) WithAliases(aliases []string) {
	n.Metadata.Aliases = append(n.Metadata.Aliases, NormalizeAliases(aliases)...)
	n.Aliases = n.Metadata.Aliases
}

// FailureModeMetadata is the metadata contract for failure mode nodes.
type FailureModeMetadata struct {
	GID           string   `json:"gid"`
	Ontology      string   `json:"ontology"`
	OntologyLayer string   `json:"ontology_layer"`
	Aliases       []string `json:"aliases,omitempty"`
	Description   string   `json:"description"`
	HandledBy     []string `json:"handled_by,omitempty"`
	Source        *string  `json:"source,omitempty"`
	Stale         bool     `json:"stale"`
	StaleReason   *string  `json:"stale_reason,omitempty"`
	LastSeenAt    *string  `json:"last_seen_at,omitempty"`
}

// FailureModeNode is a failure mode node for graph insertion.
type FailureModeNode struct {
	GID         string              `json:"gid"`
	Name        string              `json:"name"`
	ElementType string              `json:"element_type"`
	Description string              `json:"description"`
	Env         string              `json:"env"`
	Metadata    FailureModeMetadata `json:"metadata"`
}

// NewFailureModeNode builds a failure mode node (parity with
// FailureModeNode::new).
func NewFailureModeNode(env, scope, failureID, name, description string) FailureModeNode {
	g := NewGid(env, scope, TypeFailureMode, failureID).Format()
	seed := []string{NormalizeAlias(name)}
	return FailureModeNode{
		GID:         g,
		Name:        name,
		ElementType: TypeFailureMode,
		Description: description,
		Env:         env,
		Metadata: FailureModeMetadata{
			GID:           g,
			Ontology:      "procedural",
			OntologyLayer: "procedural",
			Aliases:       seed,
			Description:   description,
		},
	}
}

// elementFilePrefix marks ontology rows in code_elements.file_path, exactly
// like the Rust engine ("ontology://<gid>").
const elementFilePrefix = "ontology://"

// IsOntologyFilePath reports whether a stored element file_path belongs to
// the ontology layer.
func IsOntologyFilePath(path string) bool {
	return strings.HasPrefix(path, elementFilePrefix)
}

// toMap converts a metadata struct into the map form store.Element carries.
func toMap(v any) map[string]any {
	b, err := json.Marshal(v)
	if err != nil {
		return map[string]any{}
	}
	var m map[string]any
	if err := json.Unmarshal(b, &m); err != nil {
		return map[string]any{}
	}
	return m
}

// jsonStrArray reads a []string field out of a metadata map, tolerating a
// missing or non-array field (parity with json_str_array).
func jsonStrArray(meta map[string]any, key string) []string {
	raw, ok := meta[key]
	if !ok {
		return nil
	}
	arr, ok := raw.([]any)
	if !ok {
		return nil
	}
	out := make([]string, 0, len(arr))
	for _, v := range arr {
		if s, ok := v.(string); ok {
			out = append(out, s)
		}
	}
	return out
}
